### ISSUE TITLE
Star.exe hangs on during star_loop_8_gwworkers_test on a certain machine 

### ISSUE THREAD

**Issue by alemoi** [Starcounter/Starcounter#3566]
As you know we now have extended daily builds for both pnext and develop. star_loop_8_gwworkers_test  is always failing there (develop). The test uses 8 gateway workers and star.exe to start a codehost. Here is the test: `Level1\src\Server\sccli\star_loop_8_gwworkers_test.bat`

On sctestsrv03 the test is always hanging during Starcounter.Server.Commands. I run it with `--logsteps` so here are the examples where the log steps are stopping:
1.
```
Starcounter.Host.Loader: Query module schema information updated.: 02.51.52.752
Starcounter.Host.Loader: System type specifications initialized.: 02.51.52.803
Starcounter.Host.Loader: CLR view meta-data were populated for the given types.: 02.51.52.990
Starcounter.Host.Loader: Internal handlers were registered.: 02.51.54.047
Starcounter.Host.Loader: Processing completed.: 02.51.54.047
Starcounter.Host.Loader: Package processed.: 02.51.54.047
Starcounter.Host: Base package loaded.: 02.51.54.047
Starcounter.Host: Start completed.: 02.51.54.047
Starcounter.Host: Booting completed.: 02.51.54.048
Starcounter.Server.Commands: End task (cancelled=False): 02.51.55.158
Starcounter.Server.Commands: Executing completed (2ccca85f-943a-4b2b-b111-489895d83dc6): 02.51.55.158
[No more messages after this]
```
2.
```
Starcounter.Host.Loader: CLR view meta-data were populated for the given types.: 03.13.59.149
Starcounter.Host.Loader: Internal handlers were registered.: 03.13.59.160
Starcounter.Host.Loader: Processing completed.: 03.13.59.160
Starcounter.Host.Loader: Package processed.: 03.13.59.160
Starcounter.Host: Base package loaded.: 03.13.59.160
Starcounter.Host: Start completed.: 03.13.59.160
Starcounter.Host: Booting completed.: 03.13.59.161
Starcounter.Server.Commands: End task (cancelled=False): 03.13.59.161
Starcounter.Server.Commands: Begin task 'Restarting executables': 03.13.59.161
Starcounter.Server.Commands: End task (cancelled=True): 03.13.59.164
Starcounter.Server.Commands: Executing completed (eba278bf-86eb-466e-ac4a-09c0459afdc8): 03.13.59.167
Starcounter.Server.Commands: Executing 'Synchronizing server state for exiting code host "default", PID=9008, StartTime=5/5/2016 3:13:27 PM' (e7df0be3-b7b3-41de-a986-6ec21b7f2dd3): 03.13.59.167
Starcounter.Server.Commands: Executing completed (e7df0be3-b7b3-41de-a986-6ec21b7f2dd3): 03.13.59.168
[No more messages after this]
```
3.
```
Starcounter.Host.Loader: Processing completed.: 03.42.33.972
Starcounter.Host.Loader: Package processed.: 03.42.33.972
Starcounter.Host: Base package loaded.: 03.42.33.972
Starcounter.Host: Start completed.: 03.42.33.972
Starcounter.Host: Booting completed.: 03.42.33.972
Starcounter.Server.Commands: End task (cancelled=False): 03.42.33.972
Starcounter.Server.Commands: Begin task 'Restarting executables': 03.42.33.972
Starcounter.Server.Commands: End task (cancelled=True): 03.42.33.988
Starcounter.Server.Commands: Executing completed (4c8689b8-5111-49b9-a80f-2f6bfdfbf3cf): 03.46.55.794
Starcounter.Server.Commands: Executing 'Synchronizing server state for exiting code host "default", PID=9740, StartTime=5/5/2016 3:42:01 PM' (625916fe-bef8-460f-8d9a-335079d204a1): 03.46.55.794
Starcounter.Server.Commands: Executing completed (625916fe-bef8-460f-8d9a-335079d204a1): 03.46.55.795
[No more messages after this]
```

Here is an example of a correct boot log:
```
Starcounter.Host: Base package loaded.: 03.02.29.148
Starcounter.Host: Start completed.: 03.02.29.148
Starcounter.Host: Booting completed.: 03.02.29.148
Starcounter.Server.Commands: End task (cancelled=False): 03.02.29.148
Starcounter.Server.Commands: Begin task 'Restarting executables': 03.02.29.148
Starcounter.Server.Commands: End task (cancelled=True): 03.02.29.148
Starcounter.Server.Commands: Executing completed (16e6e421-01ff-406b-893b-57218fd24cea): 03.02.29.148
Starcounter.Server.Commands: Executing 'Synchronizing server state for exiting code host "default", PID=11432, StartTime=5/5/2016 3:02:06 PM' (bba8d68a-492e-4b3d-bccd-b466d3a8e4f1): 03.02.29.148
Starcounter.Server.Commands: Executing completed (bba8d68a-492e-4b3d-bccd-b466d3a8e4f1): 03.02.29.148
Starcounter.Server.Commands: Executing 'Starting NetworkIoTest in default' (b78c262c-266c-4b22-bdd2-bd6030d421c1): 03.02.29.210
Starcounter.Server.Commands: Begin task 'Preparing user executables and files': 03.02.29.210
Starcounter.Server.Commands: Weaving completed.: 03.02.29.366
Starcounter.Server.Commands: End task (cancelled=False): 03.02.29.366
Starcounter.Server.Commands: Begin task 'Loading executable in code host': 03.02.29.366
Starcounter.Host.Loader: Loader started.: 03.02.29.444
...
```

As you can see, more messages from `Starcounter.Server.Commands` are expected.

At this point in startup sequence, the gateway should already be configured, so seems the gateway is not the issue.

When hanging occurs the CPU for all SC processes is not used. 

The problem is not reproducible when I run the code manually :) Only when triggered in TeamCity build.

Here is the state of sccode, when I attach to it (when it hangs):
```
![image](https://cloud.githubusercontent.com/assets/208134/15047089/74b21c0c-12e3-11e6-9343-2ec4fe651816.png)

```
same for scadminserver:
```
![image](https://cloud.githubusercontent.com/assets/208134/15047101/848b6516-12e3-11e6-9192-1fb46f037006.png)

```

**Comment by alemoi** [comment]
So according to screenshots there only ReadFileNative that brings my attention. I will try to get more info on this.
@per-samuelsson do you have any idea what can happen within these `Starcounter.Server.Commands` calls?

**Comment by per-samuelsson** [comment]
> @per-samuelsson do you have any idea what can happen within these Starcounter.Server.Commands calls?

Not instantly, no. :cry: But it looks like the admin server somehow hangs, according to those logs. If you want me to step in and try to track it down, let me know! I guess my initial thing I would do is to try to attach to admin server to see if I could get some info from there. But it's hard when it's not easily reproducible.

**Comment by alemoi** [comment]
I got a suggestion from @linkdata that it might be related to console output(IO as stack trace also suggests), since it happens only when run by TeamCity. But then I wonder why its only this test and only with special configuration, 8 gw workers...

**Comment by per-samuelsson** [comment]
> I got a suggestion from @linkdata that it might be related to console output(IO as stack trace also suggests), since it happens only when run by TeamCity.

That's a good suggestion. But then again, this is the admin server - how much does it write to the console? Maybe it something with that routing of that? I'm not sure the details of console routing, but maybe it redirect the code host streams and something there make it fail?

Is it constantly reproducible on that server, @alemoi? If so, maybe we can add more diagnostics and/or disable console routing to see if it stop appearing?

**Comment by per-samuelsson** [comment]
I've studied the admin server code a bit, and I think it's likely that the hang occur during execution of [CommandProcessor.SetCompleted()](https://github.com/Starcounter/Starcounter/blob/57c6d056c5a43e4ffabbea3f261e3f30219220b1/src/Server/Starcounter.Server/Commands/CommandProcessor.cs#L200). Probably due to some code that react to the setting of that event by doing something that mess with command dispatching.

So one idea is that we add fine-grained diagnostic traces between calls there to get closer to where it actually hangs.

**Comment by alemoi** [comment]
Yes, its constantly reproducible on daily extended. One should just enable this test and disable others. While thinking about it, I am actually thinking to have an env variable to disable all tests but specified in this env var separated by commas :) Will add this to the build system, when back on Tuesday! Yes, some extra diagnostics would be helpful. Right now, as you can see in the batch, I enabled logsteps and write to file.

**Comment by linkdata** [comment]
> That's a good suggestion. But then again, this is the admin server - how much does it write to the console? Maybe it something with that routing of that? I'm not sure the details of console routing, but maybe it redirect the code host streams and something there make it fail?

Redirecting standard I/O changes the execution patterns of the redirected application, as the parent application doing the redirection will usually have different buffer sizes, response times and locking patterns from the operating systems console subsystem.

**Comment by per-samuelsson** [comment]
>  [...] I am actually thinking to have an env variable to disable all tests but specified in this env var separated by commas.

I've had this idea too, from time to time, and wished it was there. Would be really great to be able to just white list the tests you want to run, starting from 0.

**Comment by per-samuelsson** [comment]
@alemoi 

I'll take this one on next week, assigning it to me. It's not dependent on that you do that suggested enhancement to the build system (although I really would like it :sunglasses:)).

**Comment by per-samuelsson** [comment]
I've worked this through today, trying different angles, and running special builds on that particular server, and I still have nothing concrete to report.

However, I do have a suspicion, and it relates to one thing I found that I think we must rewrite no matter what. And it could even be the cause of this, actually. It's a thing added to the admin server code some time in april, when the IPC monitor was disabled. Instead, there is now code that runs in the admin server, but *outside of the dispatcher*. It just reacts instantly to the event that a code host process terminates, and fire a `Http.DELETE` from within the (Windows) thread that raise the event. And throws exceptions from it if the delete fails. That's violates the architecture of the admin server (which is designed around a queue) and also can produce nasty things I think, if raising events in callback threads within a server.

I'm still trying to investigate how to redesign it, but it will take some time still.

**Comment by alemoi** [comment]
Thanks for investigating, @per-samuelsson ! Can you please point to that parts of code that you found. I wrote it when removed the IPC monitor, and should rewrite it according to your notes.

**Comment by per-samuelsson** [comment]
> Thanks for investigating, @per-samuelsson ! Can you please point to that parts of code that you found. I wrote it when removed the IPC monitor, and should rewrite it according to your notes.

We discussed this IRL and I've now redesigned those parts with the help of @alemoi's clarification and explaining. I think we got it right this time, but I'm not sure this is the single case we experience this problem. I'm still testing.

**Comment by dan31** [comment]
Guys @per-samuelsson @alemoi great that you take the issue! One question, do you address both develop and pnext while do testing and assessment (looking into pnext/2.3 shipped to Stable in June, or at least staged with more customers)?

**Comment by per-samuelsson** [comment]
> One question, do you address both develop and pnext while do testing and assessment [...]

Well, yeah, I doubt there's anything that is different in these areas between those two branches, so once I fix it in **develop**, I think it will be a matter of just merging the same fix in **pnext**.

**Comment by per-samuelsson** [comment]
This issue is really getting the best of me. 

I've now spent majority of yesterday and most of the day today trying to trace server / code host internals in so many ways I soon can't think of anything more. 

I've debugged processes on the build server too, and still I haven't been able to find one single glitch relating to admin server threading / command dispatching. All commands process in the order expected, no exceptions occur and thread synchronization and waiting show consistent and expected results.

Yet, the test constantly fail. And the server just sits there, with nothing to do really. And `star.exe` is busy waiting on a response on a socket. Until the build server decides the time is up and shuts it down.

I've done the redesign that fixes what I reported earlier (with gateway unregistration of code hosts) and  it doesn't solve it. And I've even disabled console routing completely in custom builds, and still the same results.

My next thing to test is to add tracing to how managment REST requests flow between `star.exe` and the admin server to see if that can give a clue.

**Comment by alemoi** [comment]
> Yet, the test constantly fail. And the server just sits there, with nothing to do really. And star.exe is busy waiting on a response on a socket. Until the build server decides the time is up and shuts it down.

Maybe the response is not delivered by the gateway somehow? Is it possible to see if its sent but not delivered?

**Comment by per-samuelsson** [comment]
> Maybe the response is not delivered by the gateway somehow?

Yes, that's one of my hypothesis at least. But I haven't been able to confirm it yet.

> Is it possible to see if its sent but not delivered?

I've added tracing calls to the node layer, and as far as I can tell, all requests from `star.exe` are responded to. But some seem to take an exceptionally long time, I still haven't found any reason why.

I've also switched out one asynchronous call but it didn't help either.

**Comment by per-samuelsson** [comment]
One thing I've found though is that a lot of code has been added over the last year that is kind of questionable in it's design and particularly when it comes to synchronization. A bunch of code execute in callbacks that potentially can lock threads for a really long time, and I suspect that somewhere in there, there is a deadlock.

I've also found code that is clearly not even considering concurrent access, that will surely fail one day or another, like the one I reported earlier in this thread (and that I've fixed in this branch temporarily), but also another one just now. I'll file separate issues for those though, to make sure they are fixed in live branches.

**Comment by per-samuelsson** [comment]
I'm soon out of ideas on how to push this issue further.

I've now setup an environment on the target server where I can debug and pause `star.exe`, `scadminserver.exe` and `sccode.exe` in concert. When things "hang", there simply isn't any user-code threads on the server that can explain it, and neither so in the code host (which wasn't likely either).

As I stated earlier, `star.exe` just sits there waiting for a call to the admin server to return, but the response just don't come. I've changed the call to be synchronous, but with the same result.

![image](https://cloud.githubusercontent.com/assets/2902045/15241550/0ba4b86e-18f1-11e6-99ec-d277d1271d65.png)

On the admin server, there's just 2 active threads: first the one sending usage statistics, which I can't see make any trouble. And then a thread not connected to any user code that blocks on a file read. This thread is never spawned explicitly by us (and neither using the thread pool in the admin server - because I trace all such calls).

![image](https://cloud.githubusercontent.com/assets/2902045/15241525/cd0c3fc8-18f0-11e6-9e37-a8629aba3d91.png)

**Comment by per-samuelsson** [comment]
> I've also found code that is clearly not even considering concurrent access, that will surely fail one day or another, like the one I reported earlier in this thread (and that I've fixed in this branch temporarily), but also another one just now. I'll file separate issues for those though, to make sure they are fixed in live branches.

#3613 and #3614

**Comment by per-samuelsson** [comment]
Still can't find anything new after what I described above. I guess my last attempt here will be to try setting up so that I can debug the gateway, and see if I can find anything there. I guess I need to trigger a debug build to do so though, to get native debugging to work.

**Comment by alemoi** [comment]
I will help you with that. We can talk on hangouts.

**Comment by per-samuelsson** [comment]
>  I guess my last attempt here will be to try setting up so that I can debug the gateway, and see if I can find anything there. I guess I need to trigger a debug build to do so though, to get native debugging to work.

Here's the internal gateway state when everything hangs. Can't really tell if it's the expected or not.

![image](https://cloud.githubusercontent.com/assets/2902045/15245501/30a489f8-1907-11e6-9b82-d560e4428ceb.png)

**Comment by per-samuelsson** [comment]
I'm leaving off now, and @alemoi is taking over. We have an environment where it's constantly reproducable on the specific server and means to debug all related processes when things hang. Let's just hope we can find what it is soon.

**Comment by per-samuelsson** [comment]
@alemoi 

Are you actively working on this? I guess not. I paused the extended builds when I was working on it, but they seem to run on every push again, constantly failing (because of this). Can't we just keep them paused while we know for sure they will always fail?

**Comment by alemoi** [comment]
Not actively. Or better I will dusable this particular test for now. Because you want to see other tests for sure.

**Comment by alemoi** [comment]
I was investigating it and found some possible reasons and added checks (that never triggered apparently), but then switched to the other task. Will get back to this once done with Linux and LLVM.

**Comment by per-samuelsson** [comment]
> Or better I will dusable this particular test for now. Because you want to see other tests for sure.

Yeah, good point. Let's just make sure we don't forget that there is still a problem underneath that we need to find and fix eventually.

> Will get back to this once done with Linux and LLVM.

Great. Let me know if I can help in any way.
